virttest/qemu_devices/qdevices.py

```diff
@@ -1616,7 +1618,16 @@ def disable(self, monitor):
         return out, ver_out
 
 
-class QVirtioFSDev(QBaseDevice):
+class QDaemonDev(QBaseDevice):
+    """
+    Virtual daemon device.
+    """
+    def clear(self):
+        """Clear virtual daemon device."""
+        raise NotImplementedError
+
```

Comment:
Well, actually you could abstract more functions here, since many functions of two children class are very similar, like `_handle_log`(log name could be a class parameter), `_daemon_alive`, `_stop_daemon` `clear` `hotplug` `verify_hotplug`

Reply:
Yes, I thought about that, but we can't decide that common method in the QDaemon now, 
Maybe a general daemon object has `start_daemon`, `stop_daemon`, `is_daemon_alive`, and `clear` common public methods while I can't confirm the specific behaviors.

For the  `hotplug`, `verify_hotplug` and `_handle_log`, it's hard to abstract the behavior since there is more difference between daemons I think, and such `_handle_log`, `_daemon_alive` and `_stop_daemon` are designed into private methods since it is not necessary to be exposed and they could be called by others public methods. And subclass had better not modify the base class private methods. So if you wanna put them into base class, we just set them into public methods.

one more, we had better abstract more universality and representativeness in the base class so that the subclass could inherit as well in the future, otherwise, the subclass needs to override the base class method every time.
And I think that the `QDaemon` class maybe just an empty class to construct a readable framework here, it is one job maybe just not modify `def cleanup_daemons(self):` when involved a new QDaemon instance so for. But we could rich it when we have more understanding of the QEMU daemons in the future.

That's just my concern.

Finally, up to you.
Thanks.

Reply:
And I think that if we didn't design a high reliability and robustness base class, will result in much more difficulty in code refactoring.
Thanks.

Reply:
> Hi @vivianQizhu
> Yes, I thought about that, but we can't decide that common method in the QDaemon now,
Emm...We might simply design it based on current two ones, anyway we have some redundant implementation now, and if we have different object structure requirement in the future, we might simply rewrite or ignore some functions in the children class, I believe that's how we utilize inheritance. Having said that, I have no strong opinion on this, just let me know your thought, thanks.

Reply:
> Emm...We might simply design it based on current two ones, anyway we have some redundant implementation now, and if 

yes, you are right, so I wanna put them in code refactoring after we rich the base class when we know how to better abstract the base class, know the base common method for a daemon model.

> we have different object structure requirement in the future, we might simply rewrite or ignore some functions in the children class, I believe that's how we utilize inheritance. Having said that, I have no strong opinion on this, just let me know your thought, thanks.

Anyway, your suggestions are also ok for me.
So follow your suggestions?

Reply:
@YongxueHong Okay, please. Thanks.

Reply:
> For the `hotplug`, `verify_hotplug` and `_handle_log`, it's hard to abstract the behavior since there is more difference between daemons I think, and such `_handle_log`, `_daemon_alive` and `_stop_daemon` are designed into private methods since it is not necessary to be exposed and they could be called by others public methods. And subclass had better not modify the base class private methods. So if you wanna put them into base class, we just set them into public methods.

BTW, I understand your concern, however, this is not really a problem from my perspective, different behaviors could be implemented in children class by rewriting them, thus I suppose it will not cause troubles as you worried, see those Bus classes in qdevices.py as reference, it will be fine :)

Reply:
> > For the `hotplug`, `verify_hotplug` and `_handle_log`, it's hard to abstract the behavior since there is more difference between daemons I think, and such `_handle_log`, `_daemon_alive` and `_stop_daemon` are designed into private methods since it is not necessary to be exposed and they could be called by others public methods. And subclass had better not modify the base class private methods. So if you wanna put them into base class, we just set them into public methods.
> 
> BTW, I understand your concern, however, this is not really a problem from my perspective, different behaviors could be implemented in children class by rewriting them, thus I suppose it will not cause troubles as you worried, see those Bus classes in qdevices.py as reference, it will be fine :)

yes, for that design model is not better, the better one is that subclass inherits the base class attributes and methods then add the new attributes and methods according to the requirement of the subclass, better not to override the base class, usually, it exposes some weakness in designing the base class if one of the attributes of the base class is overridden by the subclass, show that attribute is designed unreliability, needs to evaluate them again.

When constructing parent and child classes, it is important to keep program design in mind so that overriding does not produce unnecessary or redundant code.

Inheritance in object-oriented coding can allow for adherence to the DRY (don’t repeat yourself) principle of software development, allowing for more to be done with less code and repetition. Inheritance also compels programmers to think about how they are designing the programs they are creating to ensure that code is effective and clear. 

Although we follow that completely hardly, hope could do it as for as possible.
This just my point at designing OOP.

Thanks.

Reply:
> yes, for that design model is not better, the better one is that subclass inherits the base class attributes and methods then add the new attributes and methods according to the requirement of the subclass, better not to override the base class, usually, it exposes some weakness in designing the base class if one of the attributes of the base class is overridden by the subclass, show that attribute is designed unreliability, needs to evaluate them again.

I don't agree with this...any way just keep it at your discretion, both okay for me.

Reply:
> > yes, for that design model is not better, the better one is that subclass inherits the base class attributes and methods then add the new attributes and methods according to the requirement of the subclass, better not to override the base class, usually, it exposes some weakness in designing the base class if one of the attributes of the base class is overridden by the subclass, show that attribute is designed unreliability, needs to evaluate them again.
> 
> I don't agree with this...any way just keep it at your discretion, both okay for me.

Thanks for your understanding, I will follow your suggestions, maybe you are right, have more foresight and sagacity.

Reply:
Hi @vivianQizhu 
Have updated. please help me to review, thanks.